Improve log context for exceptions

johnswanson opened the pull request:
This adds a feature to our log context that I've used and loved in the past.

For illustration, let's say I do this:

```
(defn my-fn []
  (log/with-context {:user 123 :database 5}
    (/ 1 0)))

(try (my-fn)
     (catch Exception e (log/error e)))
```

With the current code, we'll run `my-fn`, throw an exception, then *exit* the `log/with-context` block. When we eventually log the exception, the context will be empty.

This is annoying, because when the exception was thrown, we had context that a developer thought would be useful. But just when it would be *most* useful to have that information - an uncaught exception we didn't think about bubbles up to the HTTP request logging middleware or whatever - it's thrown away.

This change essentially wraps `with-context` calls with a `try/catch` that catches any exceptions and re-throws them, annotated with the original context.

This way, when we actually log the exception, we can swap back the original context. In the above example, `{:user 123 :database 5}` would be logged with the exception.

johnswanson committed:
Improve log context for exceptions

This adds a feature to our log context that I've used and loved in the
past.

For illustration, let's say I do this:

```
(defn my-fn []
  (log/with-context {:user 123 :database 5}
    (/ 1 0)))

(try (my-fn)
     (catch Exception e (log/error e)))
```

With the current code, we'll run `my-fn`, throw an exception,
then *exit* the `log/with-context` block. When we eventually log the
exception, the context will be empty.

This is annoying, because when the exception was thrown, we had context
that a developer thought would be useful. But just when it would
be *most* useful to have that information - an uncaught exception we
didn't think about bubbles up to the HTTP request logging middleware or
whatever - it's thrown away.

This change essentially wraps `with-context` calls with a `try/catch`
that catches any exceptions and re-throws them, annotated with the
original context.

This way, when we actually log the exception, we can swap back the
original context. In the above example, `{:user 123 :database 5}` would
be logged with the exception.

escherize commented on src/metabase/util/log.clj:
```diff
@@ -48,16 +66,52 @@
          (lambdaisland.glogi/log logger# level# nil s#)
          a#))))
 
+(defn with-thread-context-fn
```
What do you think of passing this an extra arg so its even more difficult to call outside the macro? Kind of like a lock and key.

escherize commented on test/metabase/util/log_test.clj:
```diff
@@ -48,3 +48,28 @@
+    (is (= {:data "inner"}
+           (log/get-exception-data
+            (try (log/with-context {:data "outer"}
+                   (log/with-context {:data "inner"}
```
Silently overwriting this feels kind of error prone.

We don't need to fix it here, I'm not sure what's best but what do you think of either

1. combining context key collisions instead of overwriting
2. warning when there's a collision

johnswanson committed:
Move log context ex-data to special key

When the exception middleware catches an exception it does some fancy
logic about whether to just let the body be the exception message (like
"Not found") or a blob of data with the message, stacktrace, data, etc.

Including the log context in the ex-data itself caused this to get
confused (because some additional data was present in the `ex-data`) and
spit out the blob where previously we'd just presented a user-friendly
error message.

To avoid doing this accidentally in the future, I just stuck all the log
context data under a special key, `:metabase.util.log/context`, in the
ex-data, which we can explicitly ignore when deciding whether to include
the whole exception in the error response.

johnswanson committed:
Fix failing test

escherize reviewed (approved):
No worries, lgtm

sansolovyov reviewed (approved):
This is a fantastic idea!
